README.md
# Single Category Permalink

A plugin for WordPress that reduces permalinks (category or post) that include entire hierarchy of categories to just having the lowest level category.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/single-category-permalink/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress.
3. Use `%category%` as a permalink tag in the `Settings` -> `Permalinks` admin options page when defining a custom permalink structure


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/single-category-permalink/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/single-category-permalink/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/single-category-permalink/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/single-category-permalink/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 2.5.2 _(2021-09-28)_
* Change: Note compatibility through WP 5.8+
* Unit tests:
    * Change: Restructure unit test directories
        * Change: Move `phpunit/bin/` into `tests/`
        * Change: Move `phpunit/` into `tests/`
    * Change: Remove 'test-' prefix from unit test file
    * Change: In bootstrap, store path to plugin file constant
    * Change: In bootstrap, add backcompat for PHPUnit pre-v6.0

## 2.5.1 _(2021-04-30)_
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)

## 2.5 _(2020-09-23)_
* Fix: Default the `$taxonomy` argument of `category_link()` to 'category' to avoid a PHP warning/error
* Fix: Handle the possibility that `get_category()` could return `null` for an invalid category ID
* New: Add a TODO item about removing deprecated functions (which is not something I want to do just yet, hence the TODO)
* Change: Update docs for return value of `category_link()` to reflect that `WP_Error` or `null` are also possible values
* Change: Note compatibility through WP 5.5+
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Unit tests:
    * New: Add tests for `category_link()`, `post_link()`
    * New: Add `unset_permalink_structures()` to unset configured permalink structures

## 2.4.2 _(2020-06-01)_
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and remove commented-out code)
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Update URLs used in examples to be HTTPS
* Change: Unit tests: Remove unnecessary unregistering of hooks

## 2.4.1 _(2020-01-18)_
* Fix: Prevent interfering with permalinks for non-category taxonomies. Props mynameisrabby.

## 2.4 _(2020-01-17)_
* Change: Use `term_link` filter instead of the deprecated `category_link` filter. Props webgmclassics.
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add more items to the list)
* Unit tests:
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * New: Add test for default hooks
    * Fix: Rename and fix miscoded test for plugin initialization
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)
* Fix: Correct typo in GitHub URL

## 2.3 _(2019-03-18)_
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hook
* Change: Initialize plugin on 'plugins_loaded' action instead of on load
* Change: Merge `do_init()` into `init()`
* CHange: Cast return value of `c2c_single_category_redirect_status` filter as integer
* Change: Split paragraph in README.md's "Support" section into two
* Change: Note compatibility through WP 5.1+
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS

## 2.2.1 _(2018-07-21)_
* Change: Switch away from using deprecated `c2c_single_category_catlink()` internally
* Change: Move `@uses` DockBlock entries to `get_http_redirect_status()`
* New: Unit tests: Add `create_hierarchical_categories()` for creating hierarchical categories so the behavior isn't duplicated in a number of tests
* New: Add README.md
* New: Add GitHub link to readme
* Change: Minor whitespace tweaks to unit test bootstrap
* Change: Note compatibility through WP 4.9+
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Improve installation instructions
* Change: Minor readme.txt tweaks
* Change: Update copyright date (2018)

## 2.2 _(2017-02-10)_
* Fix: Replace use of deprecated (in WP 4.7) `_usort_terms_by_ID()` with `wp_list_sort()` for WP 4.7+.
* Change: Change default redirect HTTP status code from 302 (temporary redirect) to 301 (permanent redirect).
* Change: Wrap functionality in class.
    * Create class `c2c_SingleCategoryPermalink`
    * Deprecate existing functions: `c2c_single_category_catlink()`, `c2c_single_category_postlink()`, `c2c_single_category_redirect()`
    * Move deprecated function functionality to class methods: `category_link()`, `post_link()`, `template_redirect()`
    * Add method `version()` to return plugin version
* Add `get_http_redirect_status()` for getting the HTTP status code for redirects.
* Change: Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable.
* Change: Enable more error output for unit tests.
* Change: Note compatibility through WP 4.7+.
* Change: Remove support for WordPress older than 4.6 (should still work for earlier versions back to WP 1.5)
* Change: Update copyright date (2017).

## 2.1.2 _(2016-03-29)_
* New: Add LICENSE file.
* New: Add empty index.php to prevent files from being listed if web server has enabled directory listings.
* New: Add 'Text Domain' to plugin header.
* Change: Minor code reformatting.
* Change: Explicitly declare methods in unit tests as public; minor unit test doc reformatting.
* Change: Note compatibility through WP 4.5+.
* Change: Update copyright date (2016).

## 2.1.1 _(2015-02-17)_
* Reformat plugin header
* Note compatibility through WP 4.1+
* Change documentation links to wp.org to be https
* Update copyright date (2015)
* Add plugin icon

## 2.1 _(2014-01-24)_
* Add unit tests
* Minor documentation improvements
* Minor code reformatting (spacing, bracing)
* Note compatibility through WP 3.8+
* Update copyright date (2014)
* Change donate link
* Add banner

## 2.0.4
* Add check to prevent execution of code if file is directly accessed
* Note compatibility through WP 3.5+

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Single_Category_Permalink
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/single-category-permalink.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Single_Category_Permalink
 */

define( 'SINGLE_CATEGORY_PERMALINK_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/single-category-permalink.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require SINGLE_CATEGORY_PERMALINK_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/test-single-category-permalink.php
<?php

defined( 'ABSPATH' ) or die();

class Single_Category_Permalink_Test extends WP_UnitTestCase {

	public function setUp() {
		parent::setUp();
		$this->set_permalink();
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	private function set_permalink( $structure = '/%category%/%postname%/', $category_structure = '/category/%category%/' ) {
		update_option( 'permalink_structure', $structure );
		update_option( 'category_base',       $category_structure );
		$GLOBALS['wp_rewrite']->init();
		add_permastruct( 'category', ltrim( $category_structure, '/' ) );
		flush_rewrite_rules();
	}

	private function unset_permalink_structures() {
		global $wp_rewrite;

		$wp_rewrite->remove_permastruct( 'category' );
		delete_option( 'permalink_structure' );
		delete_option( 'category_base' );
		$wp_rewrite->init();
	}

	public function change_redirect_status( $status ) {
		return 302;
	}

	private function create_hierarchical_categories() {
		$cat1_id = $this->factory->category->create( array ( 'slug' => 'aaa', 'name' => 'AAA' ) );
		$cat2_id = $this->factory->category->create( array ( 'slug' => 'bbb', 'name' => 'BBB', 'parent' => $cat1_id ) );
		$cat3_id = $this->factory->category->create( array ( 'slug' => 'ccc', 'name' => 'CCC', 'parent' => $cat2_id ) );

		return array( $cat1_id, $cat2_id, $cat3_id );
	}

	//
	//
	// TESTS
	//
	//


	public function test_class_exists() {
		$this->assertTrue( class_exists( 'c2c_SingleCategoryPermalink' ) );
	}

	public function test_get_version() {
		$this->assertEquals( '2.4.2', c2c_SingleCategoryPermalink::version() );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_action( 'plugins_loaded', array( 'c2c_SingleCategoryPermalink', 'init' ) ) );
	}

	public function test_default_hooks() {
		$this->assertEquals( 10, has_filter( 'term_link',         array( 'c2c_SingleCategoryPermalink', 'category_link' ) ) );
		$this->assertEquals( 10, has_filter( 'post_link',         array( 'c2c_SingleCategoryPermalink', 'post_link' ) ) );
		$this->assertEquals( 10, has_filter( 'template_redirect', array( 'c2c_SingleCategoryPermalink', 'template_redirect' ) ) );
	}

	/* Test post permalink */

	public function test_non_hierarchical_category_not_affected() {
		$cat_id  = $this->factory->category->create( array ( 'slug' => 'aaa', 'name' => 'AAA' ) );
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat_id );

		$this->assertEquals( 'http://example.org/aaa/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_hierarchical_leaf_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat3_id );

		$this->assertEquals( 'http://example.org/ccc/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_hierarchical_root_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat1_id );

		$this->assertEquals( 'http://example.org/aaa/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_hierarchical_midlevel_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat2_id );

		$this->assertEquals( 'http://example.org/bbb/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_multiple_hierarchical_leaf_categories() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$cat4_id = $this->factory->category->create( array ( 'slug' => 'ddd', 'name' => 'DDD' ) );
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, array( $cat3_id, $cat4_id ) );

		$this->assertEquals( 'http://example.org/ccc/cat-post/', get_permalink( $post_id ) );
	}

	/* Test category permalink */

	public function test_non_category_hierarchical_taxonomy_not_affected() {
		register_taxonomy( 'color', 'post', array(
			'hierarchical' => true,
			'has_archive'  => true,
			'rewrite'      => array( 'hierarchical' => true ),
		) );

		$term1_id = $this->factory->term->create( array ( 'taxonomy' => 'color', 'slug' => 'aaa', 'name' => 'AAA' ) );
		$term2_id = $this->factory->term->create( array ( 'taxonomy' => 'color', 'slug' => 'bbb', 'name' => 'BBB', 'parent' => $term1_id ) );
		$term3_id = $this->factory->term->create( array ( 'taxonomy' => 'color', 'slug' => 'ccc', 'name' => 'CCC', 'parent' => $term2_id ) );

		$this->assertEquals( 'http://example.org/color/aaa/bbb/ccc/', get_term_link( $term3_id ) );
	}

	public function test_category_permalink_for_leaf_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/ccc/', get_category_link( $cat3_id ) );
	}

	public function test_category_permalink_for_root_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/aaa/', get_category_link( $cat1_id ) );
	}

	public function test_category_permalink_for_midlevel_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/bbb/', get_category_link( $cat2_id ) );
	}

	/*
	 * get_http_redirect_status()
	 */

	public function test_get_http_redirect_status() {
		$this->assertEquals( 301, c2c_SingleCategoryPermalink::get_http_redirect_status() );
	}

	/*
	 * Filter: c2c_single_category_redirect_status
	 */

	public function test_default_redirect_status() {
		add_filter( 'c2c_single_category_redirect_status', array( $this, 'change_redirect_status' ) );

		$this->assertEquals( 302, c2c_SingleCategoryPermalink::get_http_redirect_status() );
	}

	/*
	 * category_link()
	 */

	public function test_category_link_does_not_affect_link_for_non_category() {
		$cat_id  = $this->factory->category->create( array ( 'slug' => 'aaa', 'name' => 'AAA' ) );
		$expected = 'http://example.com/some/aaa/';

		$this->assertEquals( $expected, c2c_SingleCategoryPermalink::category_link( $expected, $cat_id, 'something' ) );
	}

	public function test_category_link_returns_wp_error_for_invalid_category_id() {
		$expected = 'http://example.com/some/aaa/';

		$this->assertTrue( is_null( c2c_SingleCategoryPermalink::category_link( $expected, 999999, 'category' ) ) );
	}

	public function test_category_link_returns_wp_error_for_empty_category_id() {
		$expected = 'http://example.com/some/aaa/';

		$this->assertTrue( is_wp_error( c2c_SingleCategoryPermalink::category_link( $expected, '', 'category' ) ) );
		$this->assertTrue( is_wp_error( c2c_SingleCategoryPermalink::category_link( $expected, null, 'category' ) ) );
		$this->assertTrue( is_wp_error( c2c_SingleCategoryPermalink::category_link( $expected, false, 'category' ) ) );
	}

	public function test_category_link_when_no_category_permastruct_defined_and_invalid_category_id() {
		$this->unset_permalink_structures();

		$this->assertEquals( 'http://example.org/?cat=999999', c2c_SingleCategoryPermalink::category_link( 'url', 999999, 'category' ) );
	}

	public function test_category_link_when_no_category_permastruct_defined_and_valid_category_id() {
		$cat_id  = $this->factory->category->create( array ( 'slug' => 'aaa', 'name' => 'AAA' ) );
		$this->unset_permalink_structures();

		$this->assertEquals( 'http://example.org/?cat=' . $cat_id, c2c_SingleCategoryPermalink::category_link( 'url', $cat_id, 'category' ) );
	}

	public function test_category_link_with_valid_category_id() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/ccc/', c2c_SingleCategoryPermalink::category_link( 'url', $cat3_id, 'category' ) );
	}

	/* TODO: Test redirect of full hierarchical category permalink (post) to shorter version (e.g. /aaa/bbb/ccc/cat-post/ -> /ccc/cat-post/) */
	/* TODO: Test redirect of full hierarchical category permalink (category) to shorter version (e.g. /category/aaa/bbb/ccc/ -> /category/ccc/) */
	/* TODO: Test appropriate post(s) returned when querying shorter category permalink (e.g. /category/ccc/) */
	/* TODO: Test appropriate post(s) returned when querying leaf category off home (e.g. '/ccc/') */

}


single-category-permalink.php
<?php
/**
 * Plugin Name: Single Category Permalink
 * Version:     2.5.2
 * Plugin URI:  https://coffee2code.com/wp-plugins/single-category-permalink/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: single-category-permalink
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Reduce permalinks (category or post) that include entire hierarchy of categories to just having the lowest level category.
 *
 * Compatible with WordPress 4.6 through 5.8+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/single-category-permalinks/
 *
 * @package Single_Category_Permalink
 * @author  Scott Reilly
 * @version 2.5.2
 */

/*
	Copyright (c) 2007-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_SingleCategoryPermalink' ) ) :

class c2c_SingleCategoryPermalink {

	/**
	 * Returns version of the plugin.
	 *
	 * @since 2.2
	 */
	public static function version() {
		return '2.5.2';
	}

	/**
	 * Initialization.
	 *
	 * @since 2.2
	 */
	public static function init() {
		// Load textdomain.
		load_plugin_textdomain( 'single-category-permalink' );

		add_filter( 'term_link',         array( __CLASS__, 'category_link' ),    10, 3 );
		add_filter( 'post_link',         array( __CLASS__, 'post_link' ),        10, 2 );
		add_filter( 'template_redirect', array( __CLASS__, 'template_redirect' )       );
	}

	/**
	 * Returns the HTTP status to use for redirects.
	 *
	 * @since 2.2
	 * @uses filter c2c_single_category_redirect_code
	 *
	 * @return string
	 */
	public static function get_http_redirect_status() {
		/**
		 * Filters the HTTP status code used for redirects.
		 *
		 * @since 2.0
		 *
		 * @param int The HTTP status code to be used for redirects. Default 301.
		 */
		return (int) apply_filters( 'c2c_single_category_redirect_status', 301 );
	}

	/**
	 * Returns category URI for a given category.
	 *
	 * If the given category is hierarchical, then this function kicks into gear to
	 * reduce a hierarchical category structure to its lowest category in the link.
	 *
	 * @param  string $catlink     The default URI for the category
	 * @param  int    $category_id The category ID
	 * @param  string $taxonomy    Taxonomy slug. Default 'category'.
	 * @return string|WP_Error|null The category URI. WP_Error if category is
	 *                              empty, null if it does not exist.
	 */
	public static function category_link( $catlink, $category_id, $taxonomy = 'category' ) {
		global $wp_rewrite;

		// Bail early if taxonomy is not 'category'.
		if ( 'category' !== $taxonomy ) {
			return $catlink;
		}

		$catlink = $wp_rewrite->get_category_permastruct();

		if ( ! $catlink ) {
			$file    = trailingslashit( get_option( 'siteurl' ) );
			$catlink = $file . '?cat=' . $category_id;
		} else {
			$category = get_category( $category_id );
			if ( ! $category || is_wp_error( $category ) ) {
				return $category;
			}
			$category_nicename = $category->slug;

			//$catlink = str_replace('/category/', '/', $catlink);
			$catlink = str_replace( '%category%', $category_nicename, $catlink );
			$catlink = home_url( user_trailingslashit( $catlink, 'category' ) );
		}

		return $catlink;
	}

	/**
	 * Returns post URI for a given post.
	 *
	 * If the post permalink structure includes %category%, then this function
	 * kicks into gear to reduce a hierarchical category structure to its lowest
	 * category.
	 *
	 * @param  string  $permalink The default URI for the post
	 * @param  WP_Post $post      The post
	 * @return string             The post URI.
	 */
	public static function post_link( $permalink, $post ) {
		$permalink_structure = get_option( 'permalink_structure' );

		// Only do anything if '%category%' is part of the post permalink
		if ( strpos( $permalink_structure, '%category%' ) !== false ) {

			// Find the canonical category for the post (assigned category with
			// lowest id)
			$cats = get_the_category( $post->ID );
			if ( $cats ) {
				// Order categories by term_id.
				if ( function_exists( 'wp_list_sort' ) ) { // Introduced in WP 4.7
					$cats = wp_list_sort( $cats, 'term_id' );
				} else {
					usort( $cats, '_usort_terms_by_ID' );
				}
				$category = $cats[0];
			} else {
				$category = get_category( absint( get_option( 'default_category' ) ) );
			}

			// Find category hierachy for the category. By default, these would be
			// part of the full category permalink.
			$category_hierarchy = $category->slug;
			if ( $parent = $category->parent ) {
				$category_hierarchy = get_category_parents( $parent, false, '/', true ) . $category->slug;
			}

			// Now that the permalink component involving category hierarchy consists of is known, get rid of it.
			$permalink = str_replace( $category_hierarchy, $category->slug, $permalink );
		}

		return $permalink;
	}

	/**
	 * Redirects fully hierarchical category links to the single category link.
	 *
	 * @since 2.0
	 */
	public static function template_redirect() {
		global $wp_query, $post;

		$redirect      = null;
		$category_name = isset( $wp_query->query['category_name'] ) ? $wp_query->query['category_name'] : '';

		if ( is_category() ) {
			if ( $category_name && $category_name != $wp_query->query_vars['category_name'] ) {
				$redirect = c2c_SingleCategoryPermalink::category_link( '', $wp_query->query_vars['cat'] );
			}
		}
		elseif ( is_single() ) {
			if ( $category_name && substr_count( $category_name, '/' ) > 1 ) {
				$redirect = get_permalink( $post );
			}
		}

		if ( $redirect ) {
			wp_redirect( $redirect, self::get_http_redirect_status() );
		}
	}

} // end c2c_SingleCategoryPermalink

add_action( 'plugins_loaded', array( 'c2c_SingleCategoryPermalink', 'init' ) );

endif; // end if !class_exists()


if ( ! function_exists( 'c2c_single_category_postlink' ) ) :
/**
 * Returns post URI for a given post.
 *
 * If the post permalink structure includes %category%, then this function
 * kicks into gear to reduce a hierarchical category structure to its lowest
 * category.
 *
 * @deprecated 2.2 Use c2c_SingleCategoryPermalink::post_link() instead.
 *
 * @param  string  $permalink The default URI for the post
 * @param  WP_Post $post      The post
 * @return string  The post URI
 */
function c2c_single_category_postlink( $permalink, $post ) {
	_deprecated_function( __FUNCTION__, '2.2', 'c2c_SingleCategoryPermalink::post_link()' );

	return c2c_SingleCategoryPermalink::post_link( $permalink, $post );
}
endif;

if ( ! function_exists( 'c2c_single_category_catlink' ) ) :
/**
 * Returns category URI for a given category.
 *
 * If the given category is hierarchical, then this function kicks into gear to
 * reduce a hierarchical category structure to its lowest category in the link.
 *
 * @deprecated 2.2 Use c2c_SingleCategoryPermalink::category_link() instead.
 *
 * @param  string $catlink     The default URI for the category
 * @param  int    $category_id The category ID
 * @return string The category URI
 */
function c2c_single_category_catlink( $catlink, $category_id ) {
	_deprecated_function( __FUNCTION__, '2.2', 'c2c_SingleCategoryPermalink::category_link()' );

	return c2c_SingleCategoryPermalink::category_link( $catlink, $category_id );
}
endif;

if ( ! function_exists( 'c2c_single_category_redirect' ) ) :
/**
 * Redirects fully hierarchical category links to the single category link.
 *
 * @since 2.0
 * @deprecated 2.2 Use c2c_SingleCategoryPermalink::template_redirect() instead.
 */
function c2c_single_category_redirect() {
	_deprecated_function( __FUNCTION__, '2.2', 'c2c_SingleCategoryPermalink::template_redirect()' );

	return c2c_SingleCategoryPermalink::template_redirect();
}
endif;
